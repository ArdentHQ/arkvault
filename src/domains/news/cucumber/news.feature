Feature: News Page

	@news-displayNews
	Scenario: Successfully display News Feed
	Given Alice is on the news page
	Then the news feed should be displayed
	
	@news-paginateNews
	Scenario: Use pagination on the news page
		Given Alice is on the news page
		When she uses pagination to view the next page of articles
		Then the next page of articles are displayed
		When she uses pagination to view the previous page of news
		Then the previous page of articles are displayed

	@news-filterNews
	Scenario: Filter the news results
		Given Alice is on the news page
		When she selects specific categories
		And searches for a specifc term
		Then the results should be filtered by category
		And the results should be filtered by the search term
		When she deselects a network
		Then the page should display that no results have been found
	
	@news-searchNoResults
	Scenario: Display no results page from search
		Given Alice is on the news page
		When she searches for a term that finds 0 results
		Then the page should display that no results have been found

