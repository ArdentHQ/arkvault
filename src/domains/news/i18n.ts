export const translations = {
	AD: "Ad",
	ADD_ASSETS: {
		ALL_ASSETS: "All Cryptoassets",
		DESCRIPTION: "Select which assets you would like to follow",
		PLACEHOLDER: "Enter the asset name",
		SELECTIONS: "Your Selections",
		TITLE: "Add Assets",
	},

	CATEGORIES: {
		ALL: "All",
		COMMUNITY: "Community",
		EMERGENCY: "Emergency",
		MARKETING: "Marketing",
		TECHNICAL: "Technical",
	},

	NEWS: "News",

	NEWS_OPTIONS: {
		CATEGORY_WARNING: "At least one category needs to be selected",
		FILTER_ASSETS: "Filter Assets",
		PLACEHOLDER: "Search",
		SELECT_YOUR_CATEGORIES: "Select your categories",
		UPDATE_FILTER: "Update Filter",
		YOUR_CURRENT_SELECTIONS: "Your current selections",
	},

	PAGE_NEWS: {
		ERRORS: {
			NETWORK_ERROR: "Failed to retrieve news items. Please retry in a few minutes",
		},
		POWERED_BY: "Powered by",
		TITLE: "News",
	},
};
