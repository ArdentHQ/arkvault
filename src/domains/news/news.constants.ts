export const AVAILABLE_CATEGORIES = ["Community", "Emergency", "Marketing", "Technical"] as const;
