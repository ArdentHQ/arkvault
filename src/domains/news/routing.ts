import { ProfilePaths } from "@/router/paths";
import { RouteItem } from "@/router/router.types";
import preloadLazy from "@/utils/preload-lazy";

const News = preloadLazy(() => import("./pages/News"));

export const NewsRoutes: RouteItem[] = [
	{
		component: News,
		exact: true,
		path: ProfilePaths.News,
	},
];
