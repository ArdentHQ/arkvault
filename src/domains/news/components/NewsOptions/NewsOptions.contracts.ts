import { AvailableNewsCategories, NewsFilters } from "@/domains/news/news.contracts";

interface Option {
	name: AvailableNewsCategories;
	isSelected: boolean;
}

interface NewsOptionsProperties {
	selectedCategories: AvailableNewsCategories[];
	selectedCoins: string[];
	onSubmit: (data: NewsFilters) => void;
}

export type { NewsOptionsProperties, Option };
