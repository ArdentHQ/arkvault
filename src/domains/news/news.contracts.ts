import { AVAILABLE_CATEGORIES } from "./news.constants";

interface NewsFilters {
	categories: AvailableNewsCategories[];
	coins: string[];
	searchQuery?: string;
}

interface NewsQuery {
	categories?: string[];
	coins: string[];
	page?: number;
	query?: string;
}

export interface NewsHeaderFiltersProperties {
	coins: string[];
	categories: AvailableNewsCategories[];
	shouldResetFields: boolean;
	onSearch?: (query: string) => void;
	onFilterChange: (filters: NewsFilters) => void;
}

export type AvailableNewsCategories = typeof AVAILABLE_CATEGORIES[number] | "All";

export type { NewsFilters, NewsQuery };
