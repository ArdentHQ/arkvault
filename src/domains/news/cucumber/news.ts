import { Selector } from "testcafe";

import { buildTranslations } from "../../../app/i18n/helpers";
import { cucumber, mockRequest, visitWelcomeScreen } from "../../../utils/e2e-utils";
import { goToNews } from "../e2e/common";

const itemsPerPage = 15;
const translations = buildTranslations();
const searchBarButton = Selector('[data-testid="HeaderSearchBar"] button');
const queryInput = Selector('[data-testid="HeaderSearchBar__input"] input');
const newsCard = Selector('[data-testid="NewsCard"]');

const filteredNews = "news/filtered";

const preSteps = {
	"Given Alice is on the news page": async (t: TestController) => {
		await visitWelcomeScreen(t);
		await goToNews(t);
	},
};
cucumber(
	"@news-displayNews",
	{
		...preSteps,
		"Then the news feed should be displayed": async (t: TestController) => {
			await t.expect(newsCard.exists).ok();
			await t.expect(newsCard.count).eql(itemsPerPage);
		},
	},
	[mockRequest("https://news.payvo.com/api?coins=ARK&page=1", "news/page-1")],
);

cucumber(
	"@news-paginateNews",
	{
		...preSteps,
		"When she uses pagination to view the next page of articles": async (t: TestController) => {
			await t.hover(Selector('[data-testid="Pagination__next"]'));
			await t.click(Selector('[data-testid="Pagination__next"]'));
		},
		"Then the next page of articles are displayed": async (t: TestController) => {
			await t.expect(newsCard.exists).ok();
			await t.expect(newsCard.count).eql(itemsPerPage);
		},
		"When she uses pagination to view the previous page of news": async (t: TestController) => {
			await t.hover(Selector('[data-testid="Pagination__previous"]'));
			await t.click(Selector('[data-testid="Pagination__previous"]'));
		},
		"Then the previous page of articles are displayed": async (t: TestController) => {
			await t.expect(newsCard.exists).ok();
			await t.expect(newsCard.count).eql(itemsPerPage);
		},
	},
	[
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1", "news/page-1"),
		mockRequest("https://news.payvo.com/api?coins=ARK&page=2", "news/page-2"),
	],
);
cucumber(
	"@news-filterNews",
	{
		...preSteps,
		"When she selects specific categories": async (t: TestController) => {
			for (const category of ["Marketing", "Community", "Emergency"]) {
				await t.click(Selector(`[data-testid="NewsOptions__category-${category}"]`));
			}
		},
		"And searches for a specifc term": async (t: TestController) => {
			await t.click(searchBarButton);
			const query = "major league hacking";
			await t.typeText(queryInput, query, { replace: true });
		},
		"Then the results should be filtered by category": async (t: TestController) => {
			await t.expect(newsCard.exists).ok();
			await t
				.expect(
					Selector('[data-testid="NewsCard__category"]').withText(translations.NEWS.CATEGORIES.TECHNICAL)
						.exists,
				)
				.ok();
		},
		"And the results should be filtered by the search term": async (t: TestController) => {
			await t.expect(Selector('[data-testid="NewsCard__content"]').withText(/major league hacking/i).exists).ok();
		},
		"When she deselects a network": async (t: TestController) => {
			const ark = "NetworkOption__ark.mainnet";
			await t.click(Selector(`[data-testid="${ark}"]`));
		},
		"Then the page should display that no results have been found": async (t: TestController) => {
			await t.expect(Selector('[data-testid="EmptyResults"]').exists).ok();
		},
	},
	[
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1", "news/page-1"),
		mockRequest("https://news.payvo.com/api?coins=ARK&page=2", "news/page-2"),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK%2CETH&page=1&categories=Technical&query=major+league+hacking",
			filteredNews,
		),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK&page=1&categories=Technical%2CCommunity%2CEmergency",
			filteredNews,
		),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK&page=1&categories=Community%2CEmergency%2CTechnical",
			filteredNews,
		),
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1&categories=Emergency%2CTechnical", filteredNews),
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1&categories=Technical%2CEmergency", filteredNews),
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1&categories=Technical", filteredNews),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK&page=1&categories=Technical&query=major+league+hacking",
			filteredNews,
		),
		mockRequest("https://news.payvo.com/api?coins=ARK%2CETH&page=1&categories=Technical", filteredNews),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK&page=1&categories=Technical%2CCommunity%2CEmergency&query=major+league+hacking",
			filteredNews,
		),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK&page=1&categories=Technical%2CEmergency&query=major+league+hacking",
			filteredNews,
		),
		mockRequest(
			"https://news.payvo.com/api?coins=ARK&page=1&categories=Technical&query=major+league+hackin",
			filteredNews,
		),
	],
);
cucumber(
	"@news-searchNoResults",
	{
		...preSteps,
		"When she searches for a term that finds 0 results": async (t: TestController) => {
			await t.click(searchBarButton);
			const query = "fjdskfjdfsdjfkdsfjdsfsd";
			await t.typeText(queryInput, query, { replace: true });
		},

		"Then the page should display that no results have been found": async (t: TestController) => {
			await t.expect(Selector('[data-testid="EmptyResults"]').exists).ok();
		},
	},
	[
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1", "news/page-1"),
		mockRequest("https://news.payvo.com/api?coins=ARK&page=1&query=fjdskfjdfsdjfkdsfjdsfsd", "news/empty-response"),
	],
);
