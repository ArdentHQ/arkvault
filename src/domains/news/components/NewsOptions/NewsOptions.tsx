import { Networks } from "@payvo/sdk";
import React, { useCallback, useEffect, useMemo, useState } from "react";
import { useTranslation } from "react-i18next";

import { SelectCategory } from "./components/SelectCategory";
import { NewsOptionsProperties, Option } from "./NewsOptions.contracts";
import { Divider } from "@/app/components/Divider";
import { FilterNetwork } from "@/app/components/FilterNetwork";
import { useEnvironmentContext } from "@/app/contexts";
import { toasts } from "@/app/services";
import { AVAILABLE_CATEGORIES } from "@/domains/news/news.constants";
import { AvailableNewsCategories } from "@/domains/news/news.contracts";

// region for scrollable sidebar on small screen
const HEADER_HEIGHT = 84;
const VERTICAL_PADDING = 20 + 32;
// endregion

export const NewsOptions: React.VFC<NewsOptionsProperties> = ({ selectedCategories, selectedCoins, onSubmit }) => {
	const { env } = useEnvironmentContext();

	const { t } = useTranslation();

	const [categories, setCategories] = useState<Option[]>(
		AVAILABLE_CATEGORIES.map((name) => ({
			isSelected: selectedCategories.length === 0 || selectedCategories.includes(name),
			name,
		})),
	);

	const [coinOptions, setCoinOptions] = useState(() => {
		const coins: Record<string, { network: Networks.Network; isSelected: boolean }> = {};

		for (const network of env.availableNetworks()) {
			const coin = network.coin();

			if (!coins[coin]) {
				coins[coin] = {
					isSelected: selectedCoins.includes(coin),
					network,
				};
			}
		}

		return Object.values(coins);
	});

	const showSelectAllCategories = useMemo(
		() => categories.some((option: Option) => !option.isSelected),
		[categories],
	);

	const handleSelectCategory = (name: string) => {
		const selected = categories.filter((category: Option) => category.isSelected);

		if (selected.length === 1 && selected[0].name === name) {
			return toasts.warning(t("NEWS.NEWS_OPTIONS.CATEGORY_WARNING"));
		}

		setCategories(
			categories.map((category: Option) =>
				category.name === name
					? {
							...category,
							isSelected: !category.isSelected,
					  }
					: category,
			),
		);
	};

	const handleSelectAllCategories = () => {
		setCategories(
			categories.map((category: Option) => ({
				...category,
				isSelected: true,
			})),
		);
	};

	const handleQueryUpdate = useCallback(() => {
		const categoryNames: AvailableNewsCategories[] = [];

		for (const category of categories) {
			if (category.name !== "All" && category.isSelected) {
				categoryNames.push(category.name);
			}
		}

		const coinNames: string[] = [];

		for (const option of coinOptions) {
			if (option.isSelected) {
				coinNames.push(option.network.coin());
			}
		}

		onSubmit({
			categories: categoryNames,
			coins: coinNames,
		});
	}, [onSubmit, categories, coinOptions]);

	useEffect(() => {
		handleQueryUpdate();
	}, [handleQueryUpdate]);

	return (
		<div data-testid="NewsOptions" className="px-10 pt-7 pb-2">
			<div className="flex flex-col space-y-8">
				<div className="flex flex-col space-y-3">
					<div className="flex items-center justify-between">
						<h5 className="font-semibold">{t("COMMON.CATEGORY")}</h5>
						{showSelectAllCategories && (
							<button
								onClick={handleSelectAllCategories}
								className="-my-px border-b border-t border-dashed border-b-current border-t-transparent text-sm font-semibold text-theme-primary-600 transition-colors hover:border-b-transparent hover:text-theme-primary-700 focus:outline-none active:text-theme-primary-400"
							>
								{t("COMMON.SELECT_ALL")}
							</button>
						)}
					</div>

					<p className="text-sm text-theme-secondary-500">{t("NEWS.NEWS_OPTIONS.SELECT_YOUR_CATEGORIES")}</p>

					<div className="-mx-1 flex flex-wrap">
						{categories.map((category, index) => (
							<SelectCategory
								data-testid={`NewsOptions__category-${category.name}`}
								key={index}
								className="p-1"
								checked={category.isSelected}
								onChange={() => handleSelectCategory(category.name)}
							>
								#{t(`NEWS.CATEGORIES.${category.name.toUpperCase()}`)}
							</SelectCategory>
						))}
					</div>
				</div>

				<Divider dashed />

				<div className="flex flex-col space-y-3">
					<h5 className="font-semibold">{t("NEWS.NEWS_OPTIONS.FILTER_ASSETS")}</h5>
					<p className="text-sm text-theme-secondary-500">
						{t("NEWS.NEWS_OPTIONS.YOUR_CURRENT_SELECTIONS")}:
					</p>

					<FilterNetwork
						options={coinOptions}
						hideViewAll
						onChange={(_, networks) => setCoinOptions(networks)}
					/>
				</div>
			</div>
		</div>
	);
};

export const NewsOptionsSticky = ({ children }) => (
	<div
		data-testid="NewsOptionsSticky"
		className="sticky top-26"
		style={{ height: `calc(100vh - (${HEADER_HEIGHT}px + ${VERTICAL_PADDING}px))` }}
	>
		<div className="max-h-full overflow-y-auto rounded-lg border-2 border-theme-primary-100 bg-theme-background dark:border-theme-secondary-800">
			{children}
		</div>
	</div>
);
