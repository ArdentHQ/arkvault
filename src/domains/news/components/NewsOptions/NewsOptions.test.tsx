import userEvent from "@testing-library/user-event";
import React from "react";

import { NewsOptions } from "./NewsOptions";
import { AvailableNewsCategories } from "@/domains/news/news.contracts";
import { render, screen, waitFor } from "@/utils/testing-library";

const categories: AvailableNewsCategories[] = ["Technical"];
const coins = ["ark"];

describe("NewsOptions", () => {
	it("should render", () => {
		const { container, asFragment } = render(
			<NewsOptions selectedCategories={categories} selectedCoins={coins} onSubmit={jest.fn()} />,
		);

		expect(container).toBeInTheDocument();
		expect(asFragment()).toMatchSnapshot();
	});

	it("should select category", () => {
		render(<NewsOptions selectedCategories={categories} selectedCoins={coins} onSubmit={jest.fn()} />);

		userEvent.click(screen.getByTestId("NewsOptions__category-Technical"));
	});

	it("should select asset", () => {
		render(<NewsOptions selectedCategories={categories} selectedCoins={coins} onSubmit={jest.fn()} />);

		const arkOption = screen.getByTestId("NetworkOption__ark.mainnet");
		userEvent.click(arkOption);
	});

	it("should emit onSubmit with all selected filters", async () => {
		const onSubmit = jest.fn();

		render(<NewsOptions selectedCategories={categories} selectedCoins={coins} onSubmit={onSubmit} />);

		userEvent.click(screen.getByTestId("NewsOptions__category-Technical"));
		userEvent.click(screen.getByTestId("NetworkOption__ark.mainnet"));

		await waitFor(() =>
			expect(onSubmit).toHaveBeenCalledWith({
				categories: ["Technical"],
				coins: ["ARK"],
			}),
		);
	});
});
