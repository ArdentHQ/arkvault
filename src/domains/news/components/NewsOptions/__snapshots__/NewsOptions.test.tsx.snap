// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsOptions should render 1`] = `
<DocumentFragment>
  <div
    class="px-10 pt-7 pb-2"
    data-testid="NewsOptions"
  >
    <div
      class="flex flex-col space-y-8"
    >
      <div
        class="flex flex-col space-y-3"
      >
        <div
          class="flex items-center justify-between"
        >
          <h5
            class="font-semibold"
          >
            Category
          </h5>
          <button
            class="-my-px border-b border-t border-dashed border-b-current border-t-transparent text-sm font-semibold text-theme-primary-600 transition-colors hover:border-b-transparent hover:text-theme-primary-700 focus:outline-none active:text-theme-primary-400"
          >
            Select All
          </button>
        </div>
        <p
          class="text-sm text-theme-secondary-500"
        >
          Select your categories
        </p>
        <div
          class="-mx-1 flex flex-wrap"
        >
          <label
            class="p-1"
            data-testid="NewsOptions__category-Community"
          >
            <input
              class="css-1r71f0z select-category-input"
              type="checkbox"
              value=""
            />
            <div
              class="css-fn0s68"
            >
              #Community
            </div>
          </label>
          <label
            class="p-1"
            data-testid="NewsOptions__category-Emergency"
          >
            <input
              class="css-1r71f0z select-category-input"
              type="checkbox"
              value=""
            />
            <div
              class="css-fn0s68"
            >
              #Emergency
            </div>
          </label>
          <label
            class="p-1"
            data-testid="NewsOptions__category-Marketing"
          >
            <input
              class="css-1r71f0z select-category-input"
              type="checkbox"
              value=""
            />
            <div
              class="css-fn0s68"
            >
              #Marketing
            </div>
          </label>
          <label
            class="p-1"
            data-testid="NewsOptions__category-Technical"
          >
            <input
              checked=""
              class="css-1r71f0z select-category-input"
              type="checkbox"
              value=""
            />
            <div
              class="css-fn0s68"
            >
              #Technical
            </div>
          </label>
        </div>
      </div>
      <div
        class="border-theme-secondary-300 dark:border-theme-secondary-800 css-13w0p49"
        type="horizontal"
      />
      <div
        class="flex flex-col space-y-3"
      >
        <h5
          class="font-semibold"
        >
          Filter Assets
        </h5>
        <p
          class="text-sm text-theme-secondary-500"
        >
          Your current selections:
        </p>
        <div
          data-testid="FilterNetwork"
        >
          <ul
            class="flex flex-wrap"
            data-testid="NetworkOptions"
          >
            <li
              class="inline-block cursor-pointer pr-5 pb-5"
              data-testid="NetworkOption__ark.mainnet"
            >
              <div
                class="relative border-theme-secondary-300 text-theme-secondary-300 dark:border-theme-secondary-800 ring-theme-background duration-100 css-14jhlj1"
              >
                <div
                  class="css-v0ob3f"
                  data-testid="NetworkIcon__icon"
                  height="16"
                  width="16"
                >
                  <svg>
                    ark.svg
                  </svg>
                </div>
                <span
                  class="flex border-2 rounded-full justify-center items-center align-middle bg-theme-background border-transparent border-theme-secondary-300 dark:border-theme-secondary-800 css-19jfgr"
                >
                  <span />
                </span>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
